alibi_detect/cd/keops/tests/test_mmd_keops.py

```diff
@@ -0,0 +1,121 @@
+batch_size_permutations = [10, 1000000]
+configure_kernel_from_x_ref = [True]
+tests_mmddrift = list(product(n_features, n_enc, preprocess, n_permutations, preprocess_at_init, update_x_ref,
+                              batch_size_permutations, configure_kernel_from_x_ref))
+n_tests = len(tests_mmddrift)
+
+
+@pytest.fixture
+def mmd_params(request):
+    return tests_mmddrift[request.param]
+
+
+@pytest.mark.skipif(not has_keops, reason='Skipping since pykeops is not installed.')
+@pytest.mark.parametrize('mmd_params', list(range(n_tests)), indirect=True)
+def test_mmd(mmd_params):
+    n_features, n_enc, preprocess, n_permutations, preprocess_at_init, update_x_ref, \
+        batch_size_permutations, configure_kernel_from_x_ref = mmd_params
+
+    np.random.seed(0)
+    torch.manual_seed(0)
+
+    x_ref = np.random.randn(n * n_features).reshape(n, n_features).astype(np.float32)
+    preprocess_fn, preprocess_kwargs = preprocess
+    to_list = False
+    if hasattr(preprocess_fn, '__name__') and preprocess_fn.__name__ == 'preprocess_list':
+        if not preprocess_at_init:
+            return
+        to_list = True
+        x_ref = [_[None, :] for _ in x_ref]
+    elif isinstance(preprocess_fn, Callable) and 'layer' in list(preprocess_kwargs.keys()) \
+            and preprocess_kwargs['model'].__name__ == 'HiddenOutput':
+        model = MyModel(n_features)
+        layer = preprocess_kwargs['layer']
+        preprocess_fn = partial(preprocess_fn, model=HiddenOutput(model=model, layer=layer))
+    else:
+        preprocess_fn = None
+
+    cd = MMDDriftKeops(
+        x_ref=x_ref,
+        p_val=.05,
+        preprocess_at_init=preprocess_at_init if isinstance(preprocess_fn, Callable) else False,
+        update_x_ref=update_x_ref,
+        preprocess_fn=preprocess_fn,
+        configure_kernel_from_x_ref=configure_kernel_from_x_ref,
+        n_permutations=n_permutations,
+        batch_size_permutations=batch_size_permutations
+    )
+    x = x_ref.copy()
+    preds = cd.predict(x, return_p_val=True)
+    assert preds['data']['is_drift'] == 0 and preds['data']['p_val'] >= cd.p_val
+    if isinstance(update_x_ref, dict):
+        k = list(update_x_ref.keys())[0]
+        assert cd.n == len(x) + len(x_ref)
+        assert cd.x_ref.shape[0] == min(update_x_ref[k], len(x) + len(x_ref))
+
+    x_h1 = np.random.randn(n * n_features).reshape(n, n_features).astype(np.float32)
+    if to_list:
+        x_h1 = [_[None, :] for _ in x_h1]
+    preds = cd.predict(x_h1, return_p_val=True)
+    if preds['data']['is_drift'] == 1:
+        assert preds['data']['p_val'] < preds['data']['threshold'] == cd.p_val
+        assert preds['data']['distance'] > preds['data']['distance_threshold']
+    else:
+        assert preds['data']['p_val'] >= preds['data']['threshold'] == cd.p_val
+        assert preds['data']['distance'] <= preds['data']['distance_threshold']
+
+    # ensure the keops MMD^2 estimate matches the pytorch implementation for the same kernel
```

Comment:
Yes agreed. I had a chat expressing this exact sentiment with @jklaise, and discussed how we could modularise the types of tests in a few ways:
1. Basic tests, making sure shapes, types etc. are all fine.
2. Functionality tests, e.g. for `_mmd2` here.
3. Cross-backend tests: making sure similar functions between backends give the same result where desired.
These could share artefacts (e.g. models, input data) but be split out in very small test functions. It might be beneficial to first flesh out how we would approach this (if we want to do this) across the whole library.